### DISCUSSION TITLE
Testing output of command

### DISCUSSION THREAD

**Discussion by Warxcell** [symfony/symfony#42682] [Q&A] [answered]
Hello, I have command which outputs following text:

`'File %s wrong hash! Actual hash: %s, expected %s!',`

Got also a test for that command that verifies the output is correct one.

```
$commandTester = new \Symfony\Component\Console\Tester\CommandTester($this->command);
self::assertSame(0, $commandTester->execute([], ['decorated' => true]));

$output = $commandTester->getDisplay();

self::assertStringContainsString('File file1path wrong size! Actual size: 4 bytes, expected 5 bytes!', $output);
self::assertStringContainsString('File file1path wrong hash! Actual hash: 098f6bcd4621d373cade4e832627b4f6, expected 5a105e8b9d40e1329780d62ea2265d8a!', $output);
```

The problem is that test fails sometimes, depending on the screen of window where tests are ran, because StreamOutput adds line breaks and tabs, so output looks actually like:

```
 [ERROR] File file1path wrong hash! Actual hash: 098f6bcd4621d373cade4e832627b4f6, expected                             \n
         5a105e8b9d40e1329780d62ea2265d8a!                                                                              \n
```

What's the appropriate way to test command output in that case?

Actual PHPUnit run can be saw here: https://github.com/Warxcell/files/runs/3394071141

**Comment by ndench** [comment]
The `$commendTester->getDisplay()` takes an optional `$normalise` parameter. Looking at [the code](https://github.com/symfony/symfony/blob/bf8ecc417fb900db91978cf6ec933319f7070083/src/Symfony/Component/Console/Tester/TesterTrait.php#L49) it appears to replace `PHP_EOL` with newlines.

You could try a similar thing and just remove the `PHP_EOL` or the newlines, eg `str_replace(\PHP_EOL, '', $commandTester->getDisplay());`

**Reply by Warxcell** [reply]
it isnt that simple, i had to do it like that https://github.com/Warxcell/files/blob/5.0/tests/Command/VerifyConsistencyCommandTest.php#L75-L76, but it heels hacky.

**Comment by javiereguiluz** [comment, marked as answer]
If you set the console window width to a very large value, you won't get any new lines. You may do that by setting the `COLUMNS` env var in this test to a large number. This is possible since Symfony 3.2. See https://symfony.com/blog/new-in-symfony-3-2-console-improvements-part-2#introduced-a-new-terminal-class

**Reply by Warxcell** [reply]
@javiereguiluz thanks, but still the same result. :( Found out that it's basically capped at 120 symbols max, because of this line: https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Console/Style/SymfonyStyle.php#L52

**Reply by javiereguiluz** [reply]
Maybe you can use the same code as DomCrawler's `text()` method to manually trim all white space from `$commendTester->getDisplay()`, including the white spaces inside the string.

See https://github.com/symfony/symfony/blob/1629b5943ecbec5058e60ad4328e6717a2224cf5/src/Symfony/Component/DomCrawler/Crawler.php#L630-L632

**Reply by Warxcell** [reply]
I ended up with

```
        $output = str_replace(PHP_EOL, '', $commandTester->getDisplay());
        $output = preg_replace('/  +/', ' ', $output);
```

**Comment by Guikingone** [comment]
Hi 👋🏻 

If needed, you can display output on multiple lines using `SymfonyStyle`:

```php
<?php

// ...

$style->info([
    'First line',
    'Second line',
]);
```

This can help testing it, especially when you have "long" content to display 🙂
